src/openms/source/FORMAT/HANDLERS/MzMLHandler.cpp

```diff
@@ -4929,724 +5068,770 @@ namespace OpenMS::Internal
-          os << "\t\t\t\t\t\t" << compression_term << "\n";
-          os << "\t\t\t\t\t\t<cvParam cvRef=\"MS\" accession=\"MS:1000786\" name=\"non-standard data array\" value=\"" << array.getName() << "\" />\n";
-          writeUserParam_(os, array, 6, "/mzML/run/spectrumList/spectrum/binaryDataArrayList/binaryDataArray/cvParam/@accession", validator);
-          os << "\t\t\t\t\t\t<binary>" << encoded_string << "</binary>\n";
-          os << "\t\t\t\t\t</binaryDataArray>\n";
+            throw Exception::ConversionError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION,
+                "Compressed mode active but counter filter not available for offset calculation.");
         }
-        os << "\t\t\t\t</binaryDataArrayList>\n";
-      }
-
-      os << "\t\t\t</spectrum>\n";
+        offset = counter_ptr_->characters();
     }
-
-    template <typename ContainerT>
-    void MzMLHandler::writeContainerData_(std::ostream& os, const PeakFileOptions& pf_options_, const ContainerT& container, const String& array_type)
+    else
     {
-      // Intensity is the same for chromatograms and spectra, the second
-      // dimension is either "time" or "mz" (both of these are controlled by
-      // getMz32Bit)
-      bool is32Bit = ((array_type == "intensity" && pf_options_.getIntensity32Bit()) || pf_options_.getMz32Bit());
-      if (!is32Bit || pf_options_.getNumpressConfigurationMassTime().np_compression != MSNumpressCoder::NONE)
-      {
-        std::vector<double> data_to_encode(container.size());
-        if (array_type == "intensity")
-        {
-          for (Size p = 0; p < container.size(); ++p)
-          {
-            data_to_encode[p] = container[p].getIntensity();
-          }
-        }
-        else
-        {
-          for (Size p = 0; p < container.size(); ++p)
-          {
-            data_to_encode[p] = container[p].getPos();
-          }
-        }
-        writeBinaryDataArray_(os, pf_options_, data_to_encode, false, array_type);
-      }
-      else
-      {
-        std::vector<float> data_to_encode(container.size());
-
-        if (array_type == "intensity")
-        {
-          for (Size p = 0; p < container.size(); ++p)
-          {
-            data_to_encode[p] = container[p].getIntensity();
-          }
-        }
-        else
+        std::streampos pos = os.tellp();
+        if (pos == -1)
         {
-          for (Size p = 0; p < container.size(); ++p)
-          {
-            data_to_encode[p] = container[p].getPos();
-          }
+            throw Exception::ConversionError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION,
+                "Failed to get output stream position (uncompressed mode).");
         }
-        writeBinaryDataArray_(os, pf_options_, data_to_encode, true, array_type);
-      }
-
+        offset = static_cast<Int64>(pos);
     }
+    spectra_offsets_.emplace_back(native_id, offset + (compress_mode_ ? 0 : 3));
```

Comment:
I don't understand that `0`.
Before it was `offset + 3` and it should be same now?